### PR TITLE
Fix OPT torch.compile SDPA fallback when attention_mask is None

### LINKED ISSUE #48924: OPT training with torch.compile decomposes SDPA into Triton kernels due to attention_mask
### Summary
When training OPT with torch.compile, an attention_mask=None input is  converted into an all-ones mask for positional embeddings. This makes the model pass an explicit mask to SDPA instead of using is_causal=True, preventing the optimized FlashAttention/oneDNN backward kernel and causing a 20–30% performance regression.

### Root Cause
In `src/transformers/models/opt/modeling_opt.py`:
```python
if attention_mask is None:
    attention_mask = torch.ones(
        (inputs_embeds.shape[0], all_seq_len), dtype=torch.bool, device=inputs_embeds.device
    )

causal_mask = create_causal_mask(
    config=self.config,
    input_tensor=inputs_embeds,
    attention_mask=attention_mask,  # <- passing non-None torch.ones mask
)
```

### PR DESCRIPTION
**Pull request by gokay-ai**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48929) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48929)

Fixes #48924

OPT was filling an all-ones `attention_mask` for learned position ids, then passing that same tensor into `create_causal_mask`. Under `torch.compile`, SDPA then gets an explicit mask instead of `is_causal=True` and falls back to slower kernels.

The ones mask is only used for position ids now. `create_causal_mask` still sees `attention_mask is None`.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/transformers/contributing) and the [Pull Request](https://huggingface.co/docs/transformers/pr_checks) checks?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes according to the [guidelines](https://github.com/huggingface/transformers/tree/main/docs)?
- [x] Did you write any new necessary [tests](https://huggingface.co/docs/transformers/testing)?

## Who can review?

Text models / attention: @ArthurZucker @Cyrilvallez @vasqu

### CODE REVIEW

**Comment by Cyrilvallez** [comment]
`create_causal_mask` will create the same mask for an all 1s mask and a None mask so I'm not sure what's really the point of this PR?

**Comment by gokay-ai** [comment]
Good question — in eager mode you’re right that an all-ones 2D mask and `None` end up equivalent for `create_causal_mask` / SDPA.

The difference shows up under `torch.compile` / tracing. In `_ignore_causal_mask_sdpa`, a concrete padding tensor takes the `is_tracing(padding_mask)` path and forces a material attn mask (`return False`), so SDPA can’t use `is_causal=True`. Passing `attention_mask=None` through to `create_causal_mask` avoids that and keeps the FlashAttention / oneDNN path (that’s the regression in #48924).

This PR only fabricates the all-ones tensor for OPT’s learned position ids; the causal-mask call keeps `None` when the caller didn’t pass a mask.

**Comment by gokay-ai** [comment]
Correction on my earlier note: under `torch.compile`, `is_tracing()` is true from `is_torchdynamo_compiling()` even when the padding mask arg is `None`, so `_ignore_causal_mask_sdpa` declines the `is_causal` skip for both `None` and an all-ones mask. You’re right that this PR doesn’t change the SDPA path in current `masking_utils`.

Closing this — thanks for the catch. If #48924 still shows a real compile regression, the fix likely belongs in the shared mask/SDPA skip logic rather than OPT alone.
